### PR TITLE
fix: avoid overflows in integer division

### PR DESCRIPTION
**Pull request by guipublic**
# Description

Ensure that no overflow occurs in euclidian division.

Unfortunately the added check is not compatible with our truncation so I had to re-work it.
This could be avoided if we disallow truncation for field elements. In doing so this would:
- break modulo operation for fields
- break casting of field elements into sized integers

The first point is fine IMHO, since we already forbid field comparison because it is not consistent across field elements (0<3 but 3<p for instance). 
Indeed, modulo is also inconsistent across field elements (0%2=0 but p%2=1 for instance)

For the second point, we could change the semantic of casting and make `x as u32` becomes simply a 32 bits range check.

## Problem\*

Euclidian division was not checking for overflow. This is a big issue since it means noir is producing under-constrained code. For instance, the prime order of the noir native field is odd, so p=2k+1, which means that p%2=1. But we can also have p=0*2+0, which means that p%2=0. The latter does not overflow.

Related to #1604  

## Summary\*

We add a check in euclidian division to ensure that the result 'qb+r' does not overflow.
But this also means that euclidian division cannot be performed on unbounded integers anymore, which has impact on the truncation operation.
This is resolved by reducing the input to a bit size less than the field modulus (with some margin).
As discussed in the description, this can be avoided if we change the semantic of the casting in noir.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs [hunk 1]
```diff
@@ -608,20 +609,80 @@ impl AcirContext {
...
+                - BigUint::from(2_u32).pow(target_bit_size);
+            dbg!(toto.bits());
+            let q_max_var =
+                self.add_constant(FieldElement::from_be_bytes_reduce(&q_max_big.to_bytes_be()));
+            self.range_constrain_var(
+                quotient_high,
+                &NumericType::Unsigned { bit_size: q_max_bits },
+            )?;
+            dbg!(&q_max_big);
+            dbg!(&q_max_bits);
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs:672]: dbg

File: crates/nargo_cli/tests/test_data/brillig_not/src/main.nr [hunk 2]
```diff
@@ -2,8 +2,8 @@
 // 
 // The features being tested is not instruction on brillig
 fn main(x: Field, y : Field)  {
-    assert(false == not_operator(x as bool));
-    assert(true == not_operator(y as bool));
+    assert(false == not_operator(x == 1));
+    assert(true == not_operator(y == 1));
```
Comment (jfecher) [inline crates/nargo_cli/tests/test_data/brillig_not/src/main.nr:6]: Why were the `as bool` casts removed?

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs]: Done

File: crates/nargo_cli/tests/test_data/brillig_not/src/main.nr (hunk 2 above)
Comment (guipublic) [reply crates/nargo_cli/tests/test_data/brillig_not/src/main.nr]: The workaround in truncate_var does not work for u1, but in fact such cast was not needed in the test suite.
I believe this is OK until we can re-write this part when we have bigints.

Comment (guipublic) [reply crates/nargo_cli/tests/test_data/brillig_not/src/main.nr]: I implemented a much better alternative which does not require the workaround and works with `as bool `castings.

**Review by kevaundray — APPROVED** [review]

**Pull request merged**